gpytorch/likelihoods/likelihood.py

```diff
@@ -1,60 +1,134 @@
+            `torch.Tensor` (log probability)
         """
-        raise NotImplementedError
+        log_prob_lambda = lambda function_samples: self.forward(function_samples).log_prob(observations)
+        log_prob = self.quadrature(log_prob_lambda, function_dist)
+        return log_prob.sum(tuple(range(len(function_dist.event_shape))))
 
-    def pyro_sample_y(self, variational_dist_f, y_obs, sample_shape, name_prefix=""):
-        raise NotImplementedError
+    def marginal(self, function_dist, *params, **kwargs):
+        """
+        Computes a predictive distribution :math:`p(y*|x*)` given either a posterior
+        distribution :math:`p(f|D,x)` or a prior distribution :math:`p(f|x)` as input.
+
+        With both exact inference and variational inference, the form of
+        :math:`p(f|D,x)` or :math:`p(f|x)` should usually be Gaussian. As a result, input
+        should usually be a MultivariateNormal specified by the mean and
+        (co)variance of :math:`p(f|...)`.
+
+        Args:
+            :attr:`function_dist` (:class:`gpytorch.distributions.MultivariateNormal`)
+                Distribution for :math:`f(x)`.
+            :attr:`kwargs`
+
+        Returns
+            Distribution object (the marginal distribution, or samples from it)
+        """
+        sample_shape = torch.Size((settings.num_likelihood_samples,))
+        function_samples = function_dist.rsample(sample_shape)
+        return self.forward(function_samples)
+
+    def __call__(self, input, *params, **kwargs):
+        # Conditional
+        if torch.is_tensor(input):
+            raise RuntimeError("Likelihoods expects a MultivariateNormal input to make predictions, got a {}".format(
+                input.__class__.__name__
+            ))
+            return super().__call__(input, *params, **kwargs)
+        # Marginal
+        elif isinstance(input, MultivariateNormal):
+            return self.marginal(input, *params, **kwargs)
+        # Error
+        else:
+            raise RuntimeError(
+                "Likelihoods expects a MultivariateNormal input to make marginal predictions, or a "
+                "torch.Tensor for conditional predictions. Got a {}".format(input.__class__.__name__)
+            )
+
+
+try:
+    import pyro
+
+    class Likelihood(_Likelihood):
+        def pyro_sample_outputs(self, observations, function_dist, *params, **kwargs):
+            name_prefix = kwargs.pop("name_prefix", "")
+            with pyro.plate(name_prefix + ".output_values_plate", function_dist.batch_shape[-1], dim=-1):
+                with pyro.poutine.block():
+                    function_samples = pyro.sample(name_prefix + ".function_values", function_dist)
+                output_dist = self(function_samples, *params, **kwargs)
+                samples = pyro.sample(name_prefix + ".output_values", output_dist, obs=observations)
+                return samples
+
+except ImportError:
+    class Likelihood(_Likelihood):
+        pass
```

Comment:
```suggestion
        def pyro_sample_outputs(self, *args, **kwargs):
            raise RuntimeError("Failed to import pyro. Is it installed correctly?")
```
More useful than the `AttributeError` this would raise otherwise.